network/phonebook/phonebook.go

```diff
@@ -29,23 +29,54 @@ import (
 // of how many addresses the phonebook actually has. ( with the retry-after logic applied )
 const getAllAddresses = math.MaxInt32
 
-// PhoneBookEntryRoles defines the roles that a single entry on the phonebook can take.
+// Roles defines the roles that a single entry on the phonebook can take.
 // currently, we have two roles : relay role and archival role, which are mutually exclusive.
 //
-//msgp:ignore PhoneBookEntryRoles
-type PhoneBookEntryRoles int
+//msgp:ignore Roles
+type Roles struct {
+	roles role
+	_     func() // func is not comparable so that Roles. This is to prevent roles misuse and direct comparison.
+}
+
+var (
+	// PhoneBookEntryRelayRole used for all the relays that are provided either via the algobootstrap SRV record
+	// or via a configuration file.
+	PhoneBookEntryRelayRole = Roles{roles: relayRole}
+	// PhoneBookEntryArchivalRole used for all the archival nodes that are provided via the archive SRV record.
+	PhoneBookEntryArchivalRole = Roles{roles: archivalRole}
+)
+
+type role uint8
+
+const (
+	relayRole role = 1 << iota
+	archivalRole
+)
+
+// Has checks if the role also has the other role
+func (r Roles) Has(other Roles) bool {
+	return r.roles&other.roles != 0
+}
 
-// PhoneBookEntryRelayRole used for all the relays that are provided either via the algobootstrap SRV record
-// or via a configuration file.
-const PhoneBookEntryRelayRole = 1
+// Is checks if the role is exactly the other role
+func (r Roles) Is(other Roles) bool {
+	return r.roles == other.roles
+}
 
-// PhoneBookEntryArchivalRole used for all the archival nodes that are provided via the archive SRV record.
-const PhoneBookEntryArchivalRole = 2
+// Assign adds the other role to the role
+func (r *Roles) Assign(other Roles) {
+	r.roles |= other.roles
+}
+
+// Remove removes the other role from the role
+func (r *Roles) Remove(other Roles) {
+	r.roles &= ^other.roles
+}
 
 // Phonebook stores or looks up addresses of nodes we might contact
 type Phonebook interface {
 	// GetAddresses(N) returns up to N addresses, but may return fewer
-	GetAddresses(n int, role PhoneBookEntryRoles) []string
+	GetAddresses(n int, role Roles) []string
```

Comment:
Now that the argument is a set of roles, it's unclear whether this returns addresses that have _any_ of these roles, or addresses that have _all_ of the given roles.